packages/integrations/vercel/test/fixtures/middleware/astro.config.mjs

```diff
@@ -2,6 +2,12 @@ import {defineConfig} from "astro/config";
 import vercel from "@astrojs/vercel/serverless";
 
 export default defineConfig({
-    adapter: vercel(),
+    adapter: vercel({
+        createLocals: ({request}) => {
+            console.log(request);
+            return {
+                "foo": "bar"
+            }
+        }}),
```

Comment:
I don't quite understand the `createLocals` option here. Is there a use case for this? Could the user initialize their locals in  `src/middleware.js` instead?

My main concern is that it's being serialized into the generated code like:

```js
	const localsFn = createLocals
 		? `(${createLocals})({ request, context })`
 		: '(() => undefined)({ request, context })';
```

And that means `createLocals` need to be self-contained (The function can't access anything outside of it).

Reply:
These are excellent questions! 

> Could the user initialize their locals in src/middleware.js instead?

No, that's because of how the architecture of this feature works, which I plan to document in the `README.md` right before the end.

[Vercel sets the architecture of the middleware](https://vercel.com/_next/image?url=%2Fstatic%2Fdocs%2Fconcepts%2Ffunctions%2Fedge-middleware-dark.png&w=1200&q=75&dpl=dpl_7X9G6SyKxrcDHCME4XbfLGf79bmF).


The Vercel middleware will always run **first**, even before Astro, and there's no way to change the order. The vercel middleware will be the **first** to receive the `Request` and the **last** to receive `Response`

> Is there a use case for this? 

Based on the constraints of the architecture, I figured out the following:
- vercel edge middleware receives a context and a request, and maybe the user wants to do something with them;
- maybe the user wants to send down to Astro some information, and this info will be used to inside the Astro page;
- maybe the user wants to run some geolocation info, and pass down the Astro page info about the localisation;

> And that means createLocals need to be self-contained (The function can't access anything outside of it).

Yes, that's the most significant constraint. Unfortunately, I couldn't find a more innovative way to do so.

Reply:
Interesting, thanks for the explanation! I think the most robust way is to introduce a new special file for this, e.g. `src/vercel-edge-middleware.js` and then the esbuild build will import that file and bundle the exported handler. A bit like Vue's [`appEntrypoint` pattern](https://docs.astro.build/en/guides/integrations-guide/vue/#appentrypoint).

It might be better in the long run 🤔